docs/smart-contracts/best-practices.md

```diff
@@ -0,0 +1,276 @@
+In addition, most modules define a governance capability as the module owner.
+Conceptually, capabilities aren't difficult to comprehend or implement.
+However, if they aren't used correctly, capabilities can make your contract vulnerable to unexpected behavior or to be exploited.
+
+The following examples demonstrate patterns and outcomes for governance and basic capabilities.
+
+### Governance
+
+The code defined for the GOVERNANCE capability determines the ownership of a module and its administrative privileges.
+For simplicity in code examples, the body of the GOVERNANCE capability is often set to true. 
+For example:
+
+```pact
+(namespace 'free)
+(module YODA GOV
+   (defcap GOV () true) ; Anyone can take over the governance of the module
+   ...
+)
+```
+
+If the code associated with a capability—like GOVERNANCE in this example—is simply set to true, no conditions are being enforced to restrict access and the capability is always granted.
+If you don't enforce any restrictions on the module administrative privileges, anyone can take control of the contract and modify its tables and functions. 
+This vulnerability might seem insignificant while you're testing in a development environment.
+However, it's important to plan for and implement access controls that prevent unauthorized use of functions that perform any type of sensitive or privileged operation.
+
+The following examples demonstrate common patterns to avoid and follow to help you make contract operations more secure.
+You can find more complex and complete examples in the `coin` contract and in the `marmalade` and `spirekey` repositories.
+
+As previously noted, the following pattern is often used in sample code for simplicity:
+
+```pact
+(namespace 'free)
+(module YODA0 GOV
+  (defcap GOV () true) ; Anyone can take control of the module the true statement
+  
+  (defun hello-world:string (input:string)
+    (format "Hello {}" [input]))
+)
+```
+
+However, this pattern makes your contract vulnerable to hijacking with no protections in place to prevent unauthorized access.
+You should only use this pattern in your local development environment and in the early stages of learning to write Pact code.
+
+Another common mistake is to read a keyset or message that doesn't enforce a signing key to grant a capability as illustrated in the following example:
+
+```pact
+(namespace 'free)
+
+(module YODA1 GOV
+  (defcap GOV () 
+    (enforce-keyset (read-keyset 'hello-world))) ; You can put any value into hello-world with no enforcement
+  
+  (defun hello-world:string (input:string)
+    (format "Hello {}" [input]))
+)
+```
+
+In this example, you haven't defined a `hello-world` keyset in the `free` namespace.
+The `read-keyset` function doesn't verify that "hello-world" represents a valid keyset object or valid values.
+With this pattern, you won't see an error but no enforcement is being performed. 
```

Comment:
You can still see an error, and enforcement _is_ being performed, but it's not enforcing what you think it's enforcing. It's just enforcing ownership of an arbitrary keyset under the key `"hello world"`.

Reply:
I experimented with this a little (using an arbitrary value and an empty keyset), and didn't see an error:
```
{
    "hello-world": {
        "keys": [],
        "pred": "keys-all"
    }
}
```

Maybe not the most realistic scenario, but it works:
```
"Loaded module free.YODA1, hash Bsc9jYvcEMq5IyuTyhbmwVgY8KA6tWuxbQNe5j3qcA0"
pact>(free.YODA1.hello-world "Robert")
"Hello Robert"
```

I expected to see an error if hello-world was an arbitrary value or didn't resemble a keyset object, so I thought it might be worth noting.